contributors/design-proposals/federated-placement-policy.md

```diff
@@ -0,0 +1,264 @@
+`--admission-control=AlwaysAdmit,OPA`.
+
+### 2. Remediation of Resource Placement
+
+When policy changes or the environment in which resources are deployed changes
+(e.g. a cluster’s PCI compliance rating gets up/down-graded), resources might
+need to be moved for them to obey the placement policy. Sometimes operators may
+decide to remediate manually, but sometimes they may want Kubernetes to
+remediate automatically.
+
+To automatically reschedule resources onto desired clusters, we introduce the
+sidecar (**opa-kube-sync**) to receive notifications from OPA that indicate
+changes in placement that would bring resources back into policy compliance. The
+sidecar reacts to notifications from OPA by updating the resources in the
+federation-apiserver:
+
+![Remediation](https://docs.google.com/drawings/d/1ehuzwUXSpkOXzOUGyBW0_7jS8pKB4yRk_0YRb1X4zsY/pub?w=812&h=288)
+
+### 3. Replication of Kubernetes Resources
+
+Administrators must be able to author policies that refer to properties of
+Kubernetes resources. For example, assuming the following sample policy (in
+English):
+
+> Certain apps must be deployed on Clusters in EU zones with sufficient PCI
+> compliance.
+
+The policy definition must refer to the geographic region and PCI compliance
+rating of federated clusters. Today, the geographic region is stored as an
+attribute on the cluster resource and the PCI compliance rating is an example of
+data that may be included in a label or annotation.
+
+When the policy engine is queried for a placement decision (e.g., by the
+admission controller), it must have access to the data representing the
+federated clusters.
+
+To provide OPA with the data representing federated clusters as well as other
+potential resource types, we use a sidecar container that is deployed alongside
+OPA. The sidecar (“opa-kube-sync”) is responsible for replicating Kubernetes
+resources into OPA:
+
+![Replication](https://docs.google.com/drawings/d/1XjdgszYMDHD3hP_2ynEh_R51p7gZRoa1DBTi4yq1rc0/pub?w=812&h=288)
+
+### 4. Policy Management
+
+Policies are written in a text-based, declarative language supported by OPA. The
+policies can be loaded into the policy engine either on startup or via HTTP
+APIs.
+
+To avoid introducing additional persistent state, we propose storing policies in
+ConfigMap resources in the host cluster. The ConfigMap can be mounted into the
+policy engine’s container. The policy engine will load the policies on startup.
+
+## Open Questions
+
+1. Instead of storing the policies in ConfigMap resources in the host cluster,
+   they could also be stored in the federation-apiserver. However, this may
+   cause problems for bootstrapping the system. For example, when the admission
+   controller is enabled in the federation-apiserver, it will deny requests if
+   it cannot communicate with the policy engine (because it “fails closed” by
+   design).
+1. If policies are stored in a ConfigMap resource then the policy engine must be
```

Comment:
injection via volumemount/envvar is preferable, for minimal coupling to the kubernetes API. volumemounted configmaps will update with changes to configmap resources